cmd/sriov/main.go

```diff
@@ -0,0 +1,163 @@
+package main
+
+import (
+	"errors"
+	"fmt"
+	"runtime"
+
+	"github.com/containernetworking/cni/pkg/ipam"
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+	"github.com/containernetworking/cni/pkg/version"
+	"github.com/intel/sriov-cni/pkg/config"
+	"github.com/intel/sriov-cni/pkg/sriov"
+	"github.com/intel/sriov-cni/pkg/utils"
+	"github.com/vishvananda/netlink"
+)
+
+func init() {
+	// this ensures that main runs only on main thread (thread group leader).
+	// since namespace ops (unshare, setns) are done for a single thread, we
+	// must ensure that the goroutine does not jump from OS thread to thread
+	runtime.LockOSThread()
+}
+
+func cmdAdd(args *skel.CmdArgs) error {
+	netConf, err := config.LoadConf(args.StdinData)
+	if err != nil {
+		return fmt.Errorf("SRIOV-CNI failed to load netconf: %v", err)
+	}
+
+	netns, err := ns.GetNS(args.Netns)
+	if err != nil {
+		return fmt.Errorf("failed to open netns %q: %v", netns, err)
+	}
+	defer netns.Close()
+
+	sm := sriov.NewSriovManager()
+	if err := sm.ApplyVFConfig(netConf); err != nil {
+		return fmt.Errorf("SRIOV-CNI failed to configure VF %q", err)
+	}
+
+	// skip the IPAM allocation for the DPDK
+	var result *types.Result
+	if netConf.DPDKMode {
+		// Cache NetConf for CmdDel
+		if err = utils.SaveNetConf(args.ContainerID, config.DefaultCNIDir, args.IfName, netConf); err != nil {
+			return fmt.Errorf("error saving NetConf %q", err)
+		}
+
+		return result.Print()
+	}
+
+	if netConf.DeviceID != "" && netConf.VFID >= 0 && netConf.Master != "" {
```

Comment:
Should the ``netConf.DeviceID`` always be satisfied at this point? we check its existence when ``LoadConf``.

Reply:
I agree, this check is redundant. But does no harm to do final check before calling the setup to make sure we have all required fields. Can remove this.